PHW#3S

Xinyun Wang

### **Proposition 1: Find All Salespeople Who Achieved Sales in 2014 but Did Not Achieve Sales in 2015**

#### **Functional Specification**

**Query Name:** Salespeople with Sales in 2014 but No Sales in 2015

**Description:** This query retrieves salespeople who made sales in 2014 but made no sales in 2015, using set operators to identify these salespeople.

**Inputs:**

- `BusinessEntityID`
- `OrderDate`

**Outputs:**

- `BusinessEntityID`
- `FirstName`
- `LastName`

**Steps:**

1. Select salespeople who made sales in 2014.
2. Select salespeople who made sales in 2015.
3. Use `EXCEPT` to find salespeople who achieved sales in 2014 but not in 2015.

#### **Assumptions**

- The data is available in the `Sales.SalesOrderHeader`, `Sales.SalesPerson`, `HumanResources.Employee`, and `Person.Person` tables.
- Each salesperson has a unique `BusinessEntityID`.
- `SalesPersonID` in `Sales.SalesOrderHeader` links to the salesperson.

#### **Example Output**

| BusinessEntityID | FirstName | LastName |
| --- | --- | --- |
| 274 | Rob | Walters |
| 275 | Gail | Erickson |

In [74]:
USE AdventureWorks2019

SELECT DISTINCT 
    SP.BusinessEntityID, P.FirstName, P.LastName
FROM 
    Sales.SalesPerson AS SP
JOIN 
    HumanResources.Employee AS E ON SP.BusinessEntityID = E.BusinessEntityID
JOIN 
    Person.Person AS P ON E.BusinessEntityID = P.BusinessEntityID
JOIN 
    Sales.SalesOrderHeader AS SOH ON SP.BusinessEntityID = SOH.SalesPersonID
WHERE 
    YEAR(SOH.OrderDate) = 2014
EXCEPT

SELECT DISTINCT 
    SP.BusinessEntityID, P.FirstName, P.LastName
FROM 
    Sales.SalesPerson AS SP
JOIN 
    HumanResources.Employee AS E ON SP.BusinessEntityID = E.BusinessEntityID
JOIN 
    Person.Person AS P ON E.BusinessEntityID = P.BusinessEntityID
JOIN 
    Sales.SalesOrderHeader AS SOH ON SP.BusinessEntityID = SOH.SalesPersonID
WHERE 
    YEAR(SOH.OrderDate) = 2015;



(17 rows affected)

Total execution time: 00:00:00.074

BusinessEntityID FirstName LastName 274 Stephen Jiang 275 Michael Blythe 276 Linda Mitchell 277 Jillian Carson 278 Garrett Vargas 279 Tsvi Reiter 280 Pamela Ansman-Wolfe 281 Shu Ito 282 José Saraiva 283 David Campbell 284 Tete Mensa-Annan 285 Syed Abbas 286 Lynn Tsoflias 287 Amy Alberts 288 Rachel Valdez 289 Jae Pak 290 Ranjit Varkey Chudukatil

### **Proposition 2: Find Vendors Who Supply Components but Not Finished Goods**

#### **Functional Specification**

**Query Name:** Vendors Supplying Components but Not Finished Goods

**Description:** This query retrieves vendors who supply products in the **"Components"** category but do not supply products in the **"Finished Goods"** category. It uses the `EXCEPT` set operator to identify such vendors.

**Inputs:**

- `VendorID`
- `ProductCategory`

**Outputs:**

- `VendorID`
- `VendorName`

**Steps:**

1. Select vendors supplying products in the "Components" category.
2. Select vendors supplying products in the "Finished Goods" category.
3. Use `EXCEPT` to find vendors who supply "Components" but not "Finished Goods".

**Assumptions:**

- The data is available in the `Purchasing.Vendor`, `Purchasing.ProductVendor`, `Production.Product`, `Production.ProductSubcategory`, and `Production.ProductCategory` tables.
- Vendor names are retrieved from the `Purchasing.Vendor` table.

**Example Output**

| VendorID | VendorName |
| --- | --- |
| 1652 | Contoso Components |
| 1672 | Northwind Traders |

In [75]:
USE AdventureWorks2019

SELECT DISTINCT
    v.BusinessEntityID AS VendorID,
    v.Name AS VendorName
FROM
    Purchasing.Vendor AS v
INNER JOIN
    Purchasing.ProductVendor AS pv ON v.BusinessEntityID = pv.BusinessEntityID
INNER JOIN
    Production.Product AS p ON pv.ProductID = p.ProductID
INNER JOIN
    Production.ProductSubcategory AS ps ON p.ProductSubcategoryID = ps.ProductSubcategoryID
INNER JOIN
    Production.ProductCategory AS pc ON ps.ProductCategoryID = pc.ProductCategoryID
WHERE
    pc.Name = 'Components'

EXCEPT

SELECT DISTINCT
    v.BusinessEntityID AS VendorID,
    v.Name AS VendorName
FROM
    Purchasing.Vendor AS v
INNER JOIN
    Purchasing.ProductVendor AS pv ON v.BusinessEntityID = pv.BusinessEntityID
INNER JOIN
    Production.Product AS p ON pv.ProductID = p.ProductID
INNER JOIN
    Production.ProductSubcategory AS ps ON p.ProductSubcategoryID = ps.ProductSubcategoryID
INNER JOIN
    Production.ProductCategory AS pc ON ps.ProductCategoryID = pc.ProductCategoryID
WHERE
    pc.Name = 'Finished Goods';



(14 rows affected)

Total execution time: 00:00:00.018

VendorID VendorName 1628 Bicycle Specialists 1638 Inline Accessories 1506 Greenwood Athletic Company 1570 First Rate Bicycles 1610 Hybrid Bicycle Center 1696 Chicago City Saddles 1586 Mitchell Sports 1508 Compete Enterprises, Inc 1658 Crowley Sport 1674 Varsity Sport Co. 1542 Hill's Bicycle Service 1576 Superior Bicycles 1672 Expert Bike Co 1680 Jackson Authority

### **Proposition 3: Find Employees Who Have Worked in Either the Marketing or Sales Departments but Not Both**

#### **Functional Specification**

**Query Name:** Employees in Marketing or Sales but Not Both

**Description:** This query retrieves employees who have been assigned to either the **"Marketing"** or **"Sales"** departments but not to both. It uses the `EXCEPT` and `UNION` set operators to identify such employees.

**Inputs:**

- `BusinessEntityID`
- `DepartmentName`

**Outputs:**

- `BusinessEntityID`
- `FirstName`
- `LastName`
- `DepartmentName`

**Steps:**

1. Select employees assigned to the "Marketing" department.
2. Select employees assigned to the "Sales" department.
3. Use `UNION` to combine both sets.
4. Use `EXCEPT` and `INTERSECT` to exclude employees in both departments.

**Assumptions:**

- The data is available in the `HumanResources.EmployeeDepartmentHistory`, `HumanResources.Department`, and `Person.Person` tables.
- Employee names are retrieved from the `Person.Person` table.

**Example Output**

| BusinessEntityID | FirstName | LastName | DepartmentName |
| --- | --- | --- | --- |
| 277 | Jillian | Carson | Sales |
| 278 | José | Saraiva | Marketing |

In [76]:
USE AdventureWorks2019

SELECT
    edh.BusinessEntityID,
    p.FirstName,
    p.LastName,
    d.Name AS DepartmentName
FROM
    HumanResources.EmployeeDepartmentHistory AS edh
INNER JOIN
    HumanResources.Department AS d ON edh.DepartmentID = d.DepartmentID
INNER JOIN
    Person.Person AS p ON edh.BusinessEntityID = p.BusinessEntityID
WHERE
    d.Name = 'Marketing'

EXCEPT

SELECT
    edh.BusinessEntityID,
    p.FirstName,
    p.LastName,
    d.Name AS DepartmentName
FROM
    HumanResources.EmployeeDepartmentHistory AS edh
INNER JOIN
    HumanResources.Department AS d ON edh.DepartmentID = d.DepartmentID
INNER JOIN
    Person.Person AS p ON edh.BusinessEntityID = p.BusinessEntityID
WHERE
    d.Name = 'Marketing'
INTERSECT
SELECT
    edh.BusinessEntityID,
    p.FirstName,
    p.LastName,
    d.Name AS DepartmentName
FROM
    HumanResources.EmployeeDepartmentHistory AS edh
INNER JOIN
    HumanResources.Department AS d ON edh.DepartmentID = d.DepartmentID
INNER JOIN
    Person.Person AS p ON edh.BusinessEntityID = p.BusinessEntityID
WHERE
    d.Name = 'Sales'

UNION

SELECT
    edh.BusinessEntityID,
    p.FirstName,
    p.LastName,
    d.Name AS DepartmentName
FROM
    HumanResources.EmployeeDepartmentHistory AS edh
INNER JOIN
    HumanResources.Department AS d ON edh.DepartmentID = d.DepartmentID
INNER JOIN
    Person.Person AS p ON edh.BusinessEntityID = p.BusinessEntityID
WHERE
    d.Name = 'Sales'

EXCEPT

SELECT
    edh.BusinessEntityID,
    p.FirstName,
    p.LastName,
    d.Name AS DepartmentName
FROM
    HumanResources.EmployeeDepartmentHistory AS edh
INNER JOIN
    HumanResources.Department AS d ON edh.DepartmentID = d.DepartmentID
INNER JOIN
    Person.Person AS p ON edh.BusinessEntityID = p.BusinessEntityID
WHERE
    d.Name = 'Marketing'
INTERSECT
SELECT
    edh.BusinessEntityID,
    p.FirstName,
    p.LastName,
    d.Name AS DepartmentName
FROM
    HumanResources.EmployeeDepartmentHistory AS edh
INNER JOIN
    HumanResources.Department AS d ON edh.DepartmentID = d.DepartmentID
INNER JOIN
    Person.Person AS p ON edh.BusinessEntityID = p.BusinessEntityID
WHERE
    d.Name = 'Sales';


(28 rows affected)

Total execution time: 00:00:00.015

BusinessEntityID FirstName LastName DepartmentName 16 David Bradley Marketing 17 Kevin Brown Marketing 18 John Wood Marketing 19 Mary Dempsey Marketing 20 Wanida Benshoof Marketing 21 Terry Eminhizer Marketing 22 Sariya Harnpadoungsataya Marketing 23 Mary Gibson Marketing 24 Jill Williams Marketing 250 Sheela Word Marketing 273 Brian Welcker Sales 274 Stephen Jiang Sales 275 Michael Blythe Sales 276 Linda Mitchell Sales 277 Jillian Carson Sales 278 Garrett Vargas Sales 279 Tsvi Reiter Sales 280 Pamela Ansman-Wolfe Sales 281 Shu Ito Sales 282 José Saraiva Sales 283 David Campbell Sales 284 Tete Mensa-Annan Sales 285 Syed Abbas Sales 286 Lynn Tsoflias Sales 287 Amy Alberts Sales 288 Rachel Valdez Sales 289 Jae Pak Sales 290 Ranjit Varkey Chudukatil Sales

### **Proposition 4: Find Customers Who Made No Purchases in 2014**

#### **Functional Specification**

**Query Name:** Customers with No Purchases in 2014

**Description:** This query retrieves customers who did not make any purchases in the year 2014. It uses the `EXCEPT` set operator to identify customers without sales during that year.

**Inputs:**

- `CustomerID`
- `OrderDate`

**Outputs:**

- `CustomerID`
- `CustomerName`

**Steps:**

1. Select all customers.
2. Select customers who made purchases in 2014.
3. Use `EXCEPT` to find customers with no purchases in 2014.

**Assumptions:**

- The data is available in the `Sales.Customer`, `Sales.SalesOrderHeader`, `Person.Person`, and `Sales.Store` tables.
- Customer names are retrieved from `Person.Person` or `Sales.Store` tables.

**Example Output**

| CustomerID | CustomerName |
| --- | --- |
| 30015 | Bike Superstore |
| 30054 | Adventure Works |

In [77]:
USE AdventureWorks2019

SELECT DISTINCT
    c.CustomerID,
    ISNULL(s.Name, p.FirstName + ' ' + p.LastName) AS CustomerName
FROM
    Sales.Customer AS c
LEFT JOIN
    Person.Person AS p ON c.PersonID = p.BusinessEntityID
LEFT JOIN
    Sales.Store AS s ON c.StoreID = s.BusinessEntityID

EXCEPT

SELECT DISTINCT
    c.CustomerID,
    ISNULL(s.Name, p.FirstName + ' ' + p.LastName) AS CustomerName
FROM
    Sales.Customer AS c
INNER JOIN
    Sales.SalesOrderHeader AS soh ON c.CustomerID = soh.CustomerID
LEFT JOIN
    Person.Person AS p ON c.PersonID = p.BusinessEntityID
LEFT JOIN
    Sales.Store AS s ON c.StoreID = s.BusinessEntityID
WHERE
    YEAR(soh.OrderDate) = 2014;


(9466 rows affected)

Displaying Top 5000 rows.

Total execution time: 00:00:00.131

CustomerID CustomerName 29484 Next-Door Bike Store 28866 Aaron Adams 29170 Alexandra Adams 18504 Angel Adams 13280 Bailey Adams 28678 Ben Adams 29491 Clamps & Brackets Co. 19822 Carlos Adams 26598 Charles Adams 19410 Chloe Adams 25292 Eduardo Adams 17163 Edward Adams 24777 Eric Adams 18814 Evan Adams 20374 Fernando Adams 19527 Gabriel Adams 12154 Hailey Adams 24786 Haley Adams 18987 Hunter Adams 26174 Isaac Adams 28360 Isaiah Adams 22191 Jack Adams 17449 Jackson Adams 17488 Jason Adams 27311 Jenna Adams 16377 Jordan Adams 20387 Jordan Adams 16877 Katelyn Adams 17548 Kaylee Adams 14581 Kevin Adams 22517 Luis Adams 14640 Mackenzie Adams 22734 Maria Adams 20528 Mason Adams 11217 Natalie Adams 17380 Nathan Adams 16977 Noah Adams 17644 Rebecca Adams 22104 Richard Adams 16986 Sara Adams 23536 Savannah Adams 16293 Sean Adams 29285 Thomas Adams 14890 Wyatt Adams 22133 Xavier Adams 29493 New Bikes Company 18083 Osarumwense Agbonile 27366 Bob Alan 17721 Cheryl Alan 23601 Kari Alan 19332 Kelvin Alan 28210 Meghan Alan 15384 Xavier Alan 24223 Abigail Alexander 17387 Aidan Alexander 26333 Alexandria Alexander 25893 Alexia Alexander 28822 Amanda Alexander 27996 Angela Alexander 22714 Angelica Alexander 26673 Austin Alexander 26584 Benjamin Alexander 15273 Brandon Alexander 14503 Brianna Alexander 20526 Brittany Alexander 18106 Carson Alexander 15817 Chloe Alexander 28611 Christian Alexander 24651 Dakota Alexander 28286 Destiny Alexander 20319 Dylan Alexander 12172 Eduardo Alexander 28185 Eric Alexander 27276 Faith Alexander 17128 Fernando Alexander 17670 Gabrielle Alexander 11837 Haley Alexander 23280 Hannah Alexander 27555 Isabella Alexander 29504 Global Plaza 11780 Jessica Alexander 11165 Jocelyn Alexander 21412 Jordan Alexander 12272 Justin Alexander 22224 Katherine Alexander 12083 Kayla Alexander 16736 Kelly Alexander 19116 Logan Alexander 21459 Luis Alexander 29043 Luke Alexander 14447 Madison Alexander 28401 Maria Alexander 24821 Melissa Alexander 29505 Imported and Domestic Cycles 25760 Natalie Alexander 18795 Nathan Alexander 13305 Oscar Alexander 19654 Richard Alexander 14078 Ryan Alexander 16760 Sydney Alexander 11248 Tristan Alexander 20421 Victoria Alexander 23144 Zachary Alexander 20075 Aaron Allen 29135 Adam Allen 22353 Amanda Allen 25580 Amber Allen 27218 Bailey Allen 18776 Connor Allen 22260 Courtney Allen 15806 Devin Allen 28028 Elijah Allen 16805 Fernando Allen 11622 Gabriel Allen 26759 Hailey Allen 17640 Haley Allen 22261 Isaac Allen 29137 Jack Allen 26866 Jada Allen 25281 Jeremy Allen 28795 Jesse Allen 24053 Jose Allen 24120 Kaitlyn Allen 16802 Lucas Allen 11316 Luke Allen 21333 Mackenzie Allen 23729 Marcus Allen 11385 Miguel Allen 16997 Natalie Allen 19721 Noah Allen 29507 eCommerce Bikes 21787 Rebecca Allen 19542 Robert Allen 13917 Samuel Allen 19409 Savannah Allen 20183 Sophia Allen 24518 Stephanie Allen 13363 Sydney Allen 13485 Thomas Allen 24802 Timothy Allen 28804 Wyatt Allen 29510 Designated Distributors 23323 Alfredo Alonso 27609 Audrey Alonso 17061 Brandi Alonso 21830 Briana Alonso 27902 Carolyn Alonso 23422 Casey Alonso 23759 Clinton Alonso 18217 Danny Alonso 27501 Darren Alonso 26017 Derrick Alonso 28512 Dwayne Alonso 26842 Felicia Alonso 18767 Gary Alonso 20958 Gerald Alonso 22587 Gina Alonso 19964 Gloria Alonso 27792 Jacquelyn Alonso 11826 Jessie Alonso 27884 Jill Alonso 28766 Joe Alonso 13749 Julio Alonso 16427 Kendra Alonso 13591 Latasha Alonso 12323 Lawrence Alonso 24457 Meghan Alonso 24842 Monique Alonso 15830 Naomi Alonso 22142 Nelson Alonso 27960 Randall Alonso 17313 Raquel Alonso 14825 Rebekah Alonso 17800 Robyn Alonso 25540 Rodney Alonso 18094 Roy Alonso 24683 Shannon Alonso 25054 Sheila Alonso 28315 Stacy Alonso 26942 Tanya Alonso 13138 Teresa Alonso 25155 Theresa Alonso 29513 Tiny Bike Boutique 11329 Andy Alvarez 17777 Armando Alvarez 28179 Billy Alvarez 16209 Carolyn Alvarez 23410 Carrie Alvarez 28557 Casey Alvarez 15625 Clinton Alvarez 24720 Deanna Alvarez 13831 Eddie Alvarez 18297 Gloria Alvarez 

### **Proposition 5: Find Products That Have Never Been Purchased**

#### **Functional Specification**

**Query Name:** Products Never Purchased

**Description:** This query retrieves products that have never been sold. It uses the `EXCEPT` set operator to identify products without any sales records.

**Inputs:**

- `ProductID`

**Outputs:**

- `ProductID`
- `ProductName`

**Steps:**

1. Select all products from the `Production.Product` table.
2. Select products that have been sold from `Sales.SalesOrderDetail`.
3. Use `EXCEPT` to find products that have never been sold.

**Assumptions:**

- The data is available in the `Production.Product` and `Sales.SalesOrderDetail` tables.

**Example Output**

| ProductID | ProductName |
| --- | --- |
| 999 | New Product A |
| 1000 | New Product B |

In [78]:
USE AdventureWorks2019

SELECT
    p.ProductID,
    p.Name AS ProductName
FROM
    Production.Product AS p

EXCEPT

SELECT DISTINCT
    p.ProductID,
    p.Name AS ProductName
FROM
    Sales.SalesOrderDetail AS sod
INNER JOIN
    Production.Product AS p ON sod.ProductID = p.ProductID;


(238 rows affected)

Total execution time: 00:00:00.016

ProductID ProductName 1 Adjustable Race 3 BB Ball Bearing 2 Bearing Ball 316 Blade 324 Chain Stays 322 Chainring 320 Chainring Bolts 321 Chainring Nut 505 Cone-Shaped Race 323 Crown Race 504 Cup-Shaped Race 325 Decal 1 326 Decal 2 327 Down Tube 409 External Lock Washer 1 411 External Lock Washer 2 403 External Lock Washer 3 404 External Lock Washer 4 406 External Lock Washer 5 408 External Lock Washer 6 407 External Lock Washer 7 410 External Lock Washer 8 405 External Lock Washer 9 341 Flat Washer 1 343 Flat Washer 2 346 Flat Washer 3 345 Flat Washer 4 348 Flat Washer 5 342 Flat Washer 6 349 Flat Washer 7 347 Flat Washer 8 344 Flat Washer 9 350 Fork Crown 331 Fork End 332 Freewheel 351 Front Derailleur Cage 352 Front Derailleur Linkage 355 Guide Pulley 398 Handlebar Tube 399 Head Tube 847 Headlights - Dual-Beam 848 Headlights - Weatherproof 4 Headset Ball Bearings 386 Hex Nut 1 387 Hex Nut 10 388 Hex Nut 11 384 Hex Nut 12 385 Hex Nut 13 393 Hex Nut 14 394 Hex Nut 15 377 Hex Nut 16 378 Hex Nut 17 396 Hex Nut 18 397 Hex Nut 19 389 Hex Nut 2 390 Hex Nut 20 391 Hex Nut 21 382 Hex Nut 22 383 Hex Nut 23 392 Hex Nut 3 395 Hex Nut 4 375 Hex Nut 5 376 Hex Nut 6 379 Hex Nut 7 380 Hex Nut 8 381 Hex Nut 9 319 HL Crankarm 358 HL Grip Tape 401 HL Hub 746 HL Mountain Frame - Black, 46 740 HL Mountain Frame - Silver, 44 509 HL Mountain Rim 516 HL Mountain Seat Assembly 491 HL Nipple 840 HL Road Frame - Black, 52 680 HL Road Frame - Black, 58 837 HL Road Frame - Black, 62 720 HL Road Frame - Red, 52 721 HL Road Frame - Red, 56 706 HL Road Frame - Red, 58 512 HL Road Rim 519 HL Road Seat Assembly 526 HL Shell 524 HL Spindle/Axle 522 HL Touring Seat Assembly 419 Internal Lock Washer 1 418 Internal Lock Washer 10 421 Internal Lock Washer 2 412 Internal Lock Washer 3 413 Internal Lock Washer 4 415 Internal Lock Washer 5 417 Internal Lock Washer 6 416 Internal Lock Washer 7 420 Internal Lock Washer 8 414 Internal Lock Washer 9 402 Keyed Washer 317 LL Crankarm 356 LL Grip Tape 400 LL Hub 507 LL Mountain Rim 514 LL Mountain Seat Assembly 490 LL Nipple 737 LL Road Frame - Black, 48 724 LL Road Frame - Black, 62 728 LL Road Frame - Red, 58 818 LL Road Front Wheel 510 LL Road Rim 517 LL Road Seat Assembly 525 LL Shell 523 LL Spindle/Axle 901 LL Touring Frame - Yellow, 54 520 LL Touring Seat Assembly 449 Lock Nut 1 450 Lock Nut 10 451 Lock Nut 11 447 Lock Nut 12 448 Lock Nut 13 456 Lock Nut 14 457 Lock Nut 15 440 Lock Nut 16 441 Lock Nut 17 460 Lock Nut 18 459 Lock Nut 19 452 Lock Nut 2 453 Lock Nut 20 454 Lock Nut 21 445 Lock Nut 22 446 Lock Nut 23 455 Lock Nut 3 458 Lock Nut 4 438 Lock Nut 5 439 Lock Nut 6 442 Lock Nut 7 443 Lock Nut 8 444 Lock Nut 9 461 Lock Ring 469 Lock Washer 1 465 Lock Washer 10 475 Lock Washer 11 471 Lock Washer 12 467 Lock Washer 13 472 Lock Washer 2 474 Lock Washer 3 463 Lock Washer 4 464 Lock Washer 5 466 Lock Washer 6 470 Lock Washer 7 468 Lock Washer 8 473 Lock Washer 9 462 Lower Head Race 851 Men's Sports Shorts, XL 476 Metal Angle 477 Metal Bar 1 478 Metal Bar 2 480 Metal Plate 1 479 Metal Plate 2 481 Metal Plate 3 488 Metal Sheet 1 482 Metal Sheet 2 483 Metal Sheet 3 485 Metal Sheet 4 486 Metal Sheet 5 487 Metal Sheet 6 484 Metal Sheet 7 489 Metal Tread Plate 995 ML Bottom Bracket 318 ML Crankarm 803 ML Fork 357 ML Grip Tape 508 ML Mountain Rim 515 ML Mountain Seat Assembly 731 ML Road Frame - Red, 44 734 ML Road Frame - Red, 58 735 ML Road Frame - Red, 60 834 ML Road Frame-W - Yellow, 42 812 ML Road Handlebars 511 ML Road Rim 518 ML Road Seat Assembly 912 ML Road Seat/Saddle 521 ML Touring Seat Assembly 328 Mountain End Caps 845 Mountain Pump 492 Paint - Black 495 Paint - Blue 493 Paint - Red 494 Paint - Silver 496 Paint - Yellow 497 Pinch Bolt 679 Rear Derailleur Cage 506 Reflector 329 Road End Caps 528 Seat Lug 530 Seat Post 532 Seat Stays 533 Seat Tube 527 Spokes 531 Steerer 529 Stem 846 Taillights - Battery-Powered 535 Tension Pulley 361 Thin-Jam Hex Nut 1 360 Thin-Jam Hex Nut 10 374 Thin-Jam Hex Nut 11 373 Thin-Jam

### **Proposition 6: Find Products That Are Accessories or Clothing**

#### **Functional Specification**

**Query Name:** Products That Are Accessories or Clothing

**Description:** This query retrieves products that belong to either the **"Accessories"** or **"Clothing"** categories. It uses the `UNION` set operator to combine products from both categories.

**Inputs:**

- `ProductID`
- `ProductCategory`

**Outputs:**

- `ProductID`
- `ProductName`
- `CategoryName`

**Steps:**

1. Select products in the "Accessories" category.
2. Select products in the "Clothing" category.
3. Use `UNION` to combine both sets.

**Assumptions:**

- The data is available in the `Production.Product`, `Production.ProductSubcategory`, and `Production.ProductCategory` tables.

**Example Output**

| ProductID | ProductName | CategoryName |
| --- | --- | --- |
| 858 | Water Bottle - 30 oz. | Accessories |
| 937 | Long-Sleeve Logo Jersey | Clothing |

In [81]:
USE AdventureWorks2019

SELECT
    p.ProductID,
    p.Name AS ProductName,
    pc.Name AS CategoryName
FROM
    Production.Product AS p
INNER JOIN
    Production.ProductSubcategory AS ps ON p.ProductSubcategoryID = ps.ProductSubcategoryID
INNER JOIN
    Production.ProductCategory AS pc ON ps.ProductCategoryID = pc.ProductCategoryID
WHERE
    pc.Name = 'Accessories'

UNION

SELECT
    p.ProductID,
    p.Name AS ProductName,
    pc.Name AS CategoryName
FROM
    Production.Product AS p
INNER JOIN
    Production.ProductSubcategory AS ps ON p.ProductSubcategoryID = ps.ProductSubcategoryID
INNER JOIN
    Production.ProductCategory AS pc ON ps.ProductCategoryID = pc.ProductCategoryID
WHERE
    pc.Name = 'Clothing';


(64 rows affected)

Total execution time: 00:00:00.006

ProductID ProductName CategoryName 707 Sport-100 Helmet, Red Accessories 708 Sport-100 Helmet, Black Accessories 709 Mountain Bike Socks, M Clothing 710 Mountain Bike Socks, L Clothing 711 Sport-100 Helmet, Blue Accessories 712 AWC Logo Cap Clothing 713 Long-Sleeve Logo Jersey, S Clothing 714 Long-Sleeve Logo Jersey, M Clothing 715 Long-Sleeve Logo Jersey, L Clothing 716 Long-Sleeve Logo Jersey, XL Clothing 841 Men's Sports Shorts, S Clothing 842 Touring-Panniers, Large Accessories 843 Cable Lock Accessories 844 Minipump Accessories 845 Mountain Pump Accessories 846 Taillights - Battery-Powered Accessories 847 Headlights - Dual-Beam Accessories 848 Headlights - Weatherproof Accessories 849 Men's Sports Shorts, M Clothing 850 Men's Sports Shorts, L Clothing 851 Men's Sports Shorts, XL Clothing 852 Women's Tights, S Clothing 853 Women's Tights, M Clothing 854 Women's Tights, L Clothing 855 Men's Bib-Shorts, S Clothing 856 Men's Bib-Shorts, M Clothing 857 Men's Bib-Shorts, L Clothing 858 Half-Finger Gloves, S Clothing 859 Half-Finger Gloves, M Clothing 860 Half-Finger Gloves, L Clothing 861 Full-Finger Gloves, S Clothing 862 Full-Finger Gloves, M Clothing 863 Full-Finger Gloves, L Clothing 864 Classic Vest, S Clothing 865 Classic Vest, M Clothing 866 Classic Vest, L Clothing 867 Women's Mountain Shorts, S Clothing 868 Women's Mountain Shorts, M Clothing 869 Women's Mountain Shorts, L Clothing 870 Water Bottle - 30 oz. Accessories 871 Mountain Bottle Cage Accessories 872 Road Bottle Cage Accessories 873 Patch Kit/8 Patches Accessories 874 Racing Socks, M Clothing 875 Racing Socks, L Clothing 876 Hitch Rack - 4-Bike Accessories 877 Bike Wash - Dissolver Accessories 878 Fender Set - Mountain Accessories 879 All-Purpose Bike Stand Accessories 880 Hydration Pack - 70 oz. Accessories 881 Short-Sleeve Classic Jersey, S Clothing 882 Short-Sleeve Classic Jersey, M Clothing 883 Short-Sleeve Classic Jersey, L Clothing 884 Short-Sleeve Classic Jersey, XL Clothing 921 Mountain Tire Tube Accessories 922 Road Tire Tube Accessories 923 Touring Tire Tube Accessories 928 LL Mountain Tire Accessories 929 ML Mountain Tire Accessories 930 HL Mountain Tire Accessories 931 LL Road Tire Accessories 932 ML Road Tire Accessories 933 HL Road Tire Accessories 934 Touring Tire Accessories

### **Proposition 7: Find Products Sold Every Year from 2012 to 2014**

#### **Functional Specification**

**Query Name:** Products Sold Every Year from 2012 to 2014

**Description:** This query retrieves products that have been sold in every year from **2012** to **2014**. It uses the `INTERSECT` set operator to find products sold in all three years.

**Inputs:**

- `ProductID`
- `OrderDate`

**Outputs:**

- `ProductID`
- `ProductName`

**Steps:**

1. Select products sold in 2012.
2. Select products sold in 2013.
3. Select products sold in 2014.
4. Use `INTERSECT` to find products sold in all three years.

**Assumptions:**

- The data is available in the `Sales.SalesOrderHeader`, `Sales.SalesOrderDetail`, and `Production.Product` tables.

**Example Output**

| ProductID | ProductName |
| --- | --- |
| 680 | HL Road Frame - Black, 58 |
| 712 | Sport-100 Helmet, Red |

In [82]:
USE AdventureWorks2019

SELECT DISTINCT
    p.ProductID,
    p.Name AS ProductName
FROM
    Sales.SalesOrderHeader AS soh
INNER JOIN
    Sales.SalesOrderDetail AS sod ON soh.SalesOrderID = sod.SalesOrderID
INNER JOIN
    Production.Product AS p ON sod.ProductID = p.ProductID
WHERE
    YEAR(soh.OrderDate) = 2012

INTERSECT

SELECT DISTINCT
    p.ProductID,
    p.Name AS ProductName
FROM
    Sales.SalesOrderHeader AS soh
INNER JOIN
    Sales.SalesOrderDetail AS sod ON soh.SalesOrderID = sod.SalesOrderID
INNER JOIN
    Production.Product AS p ON sod.ProductID = p.ProductID
WHERE
    YEAR(soh.OrderDate) = 2013

INTERSECT

SELECT DISTINCT
    p.ProductID,
    p.Name AS ProductName
FROM
    Sales.SalesOrderHeader AS soh
INNER JOIN
    Sales.SalesOrderDetail AS sod ON soh.SalesOrderID = sod.SalesOrderID
INNER JOIN
    Production.Product AS p ON sod.ProductID = p.ProductID
WHERE
    YEAR(soh.OrderDate) = 2014;


(58 rows affected)

Total execution time: 00:00:00.146

ProductID ProductName 856 Men's Bib-Shorts, M 779 Mountain-200 Silver, 38 762 Road-650 Red, 44 716 Long-Sleeve Logo Jersey, XL 862 Full-Finger Gloves, M 722 LL Road Frame - Black, 58 793 Road-250 Black, 44 822 ML Road Frame-W - Yellow, 38 799 Road-550-W Yellow, 42 836 ML Road Frame-W - Yellow, 48 730 LL Road Frame - Red, 62 742 HL Mountain Frame - Silver, 46 736 LL Road Frame - Black, 44 782 Mountain-200 Black, 38 859 Half-Finger Gloves, M 813 HL Road Handlebars 796 Road-250 Black, 58 739 HL Mountain Frame - Silver, 42 854 Women's Tights, L 808 LL Mountain Handlebars 860 Half-Finger Gloves, L 708 Sport-100 Helmet, Black 714 Long-Sleeve Logo Jersey, M 748 HL Mountain Frame - Silver, 38 797 Road-550-W Yellow, 38 711 Sport-100 Helmet, Blue 780 Mountain-200 Silver, 42 765 Road-650 Black, 58 717 HL Road Frame - Red, 62 863 Full-Finger Gloves, L 794 Road-250 Black, 48 800 Road-550-W Yellow, 44 712 AWC Logo Cap 783 Mountain-200 Black, 42 760 Road-650 Red, 60 852 Women's Tights, S 858 Half-Finger Gloves, S 795 Road-250 Black, 52 726 LL Road Frame - Red, 48 855 Men's Bib-Shorts, S 809 ML Mountain Handlebars 763 Road-650 Red, 48 743 HL Mountain Frame - Black, 42 792 Road-250 Red, 58 835 ML Road Frame-W - Yellow, 44 781 Mountain-200 Silver, 46 718 HL Road Frame - Red, 44 810 HL Mountain Handlebars 838 HL Road Frame - Black, 44 738 LL Road Frame - Black, 52 784 Mountain-200 Black, 46 801 Road-550-W Yellow, 48 707 Sport-100 Helmet, Red 715 Long-Sleeve Logo Jersey, L 761 Road-650 Red, 62 861 Full-Finger Gloves, S 747 HL Mountain Frame - Black, 38 798 Road-550-W Yellow, 40

### **Proposition 8: Find Products Sold in 'Northwest' Territory but Not in 'Southwest' Territory**

#### **Functional Specification**

**Query Name:** Products Sold in Northwest but Not Southwest Territory

**Description:** This query retrieves products that have been sold to customers in the **"Northwest"** sales territory but have not been sold to customers in the **"Southwest"** territory. It uses the `EXCEPT` set operator to identify these products.

**Inputs:**

- `ProductID`: The unique identifier for each product.
- `TerritoryID`: The unique identifier for each sales territory.
- `TerritoryName`: The name of the sales territory.

**Outputs:**

- `ProductID`: The unique identifier for the product.
- `ProductName`: The name of the product.

**Steps:**

1. Select products sold in the "Northwest" territory.
2. Select products sold in the "Southwest" territory.
3. Use `EXCEPT` to find products sold in "Northwest" but not in "Southwest".

**Assumptions:**

- Data is available in the `Sales.SalesOrderHeader`, `Sales.SalesOrderDetail`, `Production.Product`, and `Sales.SalesTerritory` tables.
- Products have unique `ProductID`s in the `Production.Product` table.

**Example Output**

| ProductID | ProductName |
| --- | --- |
| 680 | HL Road Frame - Black, 58 |
| 712 | Sport-100 Helmet, Red |

In [83]:
USE AdventureWorks2019

SELECT DISTINCT
    p.ProductID,
    p.Name AS ProductName
FROM
    Sales.SalesOrderHeader AS soh
INNER JOIN
    Sales.SalesOrderDetail AS sod ON soh.SalesOrderID = sod.SalesOrderID
INNER JOIN
    Production.Product AS p ON sod.ProductID = p.ProductID
INNER JOIN
    Sales.SalesTerritory AS st ON soh.TerritoryID = st.TerritoryID
WHERE
    st.Name = 'Northwest'

EXCEPT


SELECT DISTINCT
    p.ProductID,
    p.Name AS ProductName
FROM
    Sales.SalesOrderHeader AS soh
INNER JOIN
    Sales.SalesOrderDetail AS sod ON soh.SalesOrderID = sod.SalesOrderID
INNER JOIN
    Production.Product AS p ON sod.ProductID = p.ProductID
INNER JOIN
    Sales.SalesTerritory AS st ON soh.TerritoryID = st.TerritoryID
WHERE
    st.Name = 'Southwest';


(2 rows affected)

Total execution time: 00:00:00.093

ProductID ProductName 744 HL Mountain Frame - Black, 44 898 LL Touring Frame - Blue, 62

### **Proposition 9: Find Vendors Who Supply Products Not Sold in 2014**

#### **Functional Specification**

**Query Name:** Vendors Supplying Products Not Sold in 2014

**Description:** This query retrieves vendors who supply products that were not sold at all in the year 2014. It uses the `EXCEPT` set operator to find vendors whose products did not have any sales in that year.

**Inputs:**

- `VendorID`
- `ProductID`
- `OrderDate`

**Outputs:**

- `VendorID`
- `VendorName`

**Steps:**

1. Select vendors and products they supply.
2. Select products that were sold in 2014.
3. Use `EXCEPT` to find vendors whose products were not sold in 2014.

**Assumptions:**

- Data is available in the `Purchasing.Vendor`, `Purchasing.ProductVendor`, `Production.Product`, and `Sales.SalesOrderDetail` tables.
- Vendor names are retrieved from the `Purchasing.Vendor` table.

**Example Output**

| VendorID | VendorName |
| --- | --- |
| 1510 | International |
| 1516 | Gardner Touring Cycles |

In [84]:
USE AdventureWorks2019;

WITH VendorProducts AS (
    SELECT DISTINCT
        v.BusinessEntityID AS VendorID,
        v.Name AS VendorName,
        p.ProductID
    FROM
        Purchasing.Vendor AS v
    INNER JOIN
        Purchasing.ProductVendor AS pv ON v.BusinessEntityID = pv.BusinessEntityID
    INNER JOIN
        Production.Product AS p ON pv.ProductID = p.ProductID
)


, ProductsSoldIn2019 AS (
    SELECT DISTINCT
        sod.ProductID
    FROM
        Sales.SalesOrderDetail AS sod
    INNER JOIN
        Sales.SalesOrderHeader AS soh ON sod.SalesOrderID = soh.SalesOrderID
    WHERE
        YEAR(soh.OrderDate) = 2014
)

SELECT DISTINCT
    vp.VendorID,
    vp.VendorName
FROM
    VendorProducts AS vp

EXCEPT

SELECT DISTINCT
    vp.VendorID,
    vp.VendorName
FROM
    VendorProducts AS vp
INNER JOIN
    ProductsSoldIn2019 AS psi ON vp.ProductID = psi.ProductID;


(61 rows affected)

Total execution time: 00:00:00.074

VendorID VendorName 1492 Australia Bike Retailer 1494 Allenson Cycles 1496 Advanced Bicycles 1500 Morgan Bike Accessories 1504 Chicago Rent-All 1510 International 1514 Training Systems 1516 Gardner Touring Cycles 1518 International Trek Center 1522 First National Sport Co. 1526 International Bicycles 1530 Comfort Road Bicycles 1534 Ready Rentals 1536 Cruger Bike Company 1540 Bergeron Off-Roads 1542 Hill's Bicycle Service 1544 Circuit Cycles 1548 Consumer Cycles 1554 WestAmerica Bicycle Co. 1556 West Junction Cycles 1560 Capital Road Cycles 1562 Norstan Bike Hut 1566 Burnett Road Warriors 1568 Custom Frames, Inc. 1572 National Bike Association 1578 Vision Cycles, Inc. 1580 Litware, Inc. 1582 Inner City Bikes 1584 Trey Research 1590 SUPERSALES INC. 1592 Lindell 1598 Continental Pro Cycles 1600 Federal Sport 1602 Beaumont Bikes 1604 Bike Satellite Inc. 1608 Sport Playground 1612 Midwest Sport, Inc. 1614 Reliance Fitness, Inc. 1616 Aurora Bike Center 1618 Metro Sport Equipment 1620 Lakewood Bicycle 1622 Speed Corporation 1624 Competition Bike Training Systems 1626 Hill Bicycle Center 1644 International Sport Assoc. 1646 Electronic Bike Repair & Supplies 1648 Wide World Importers 1650 American Bicycles and Wheels 1654 American Bikes 1656 Mountain Works 1662 Northern Bike Travel 1664 Anderson's Custom Bikes 1666 Leaf River Terrain 1678 Proseware, Inc. 1682 Premier Sport, Inc. 1686 Pro Sport Industries 1688 Wood Fitness 1690 Bloomington Multisport 1692 Carlson Specialties 1694 Compete, Inc. 1698 Business Equipment Center

### **Proposition 10: Find Products Reviewed in 2013 but Not in 2014**

#### **Functional Specification**

**Query Name:** Products Reviewed in 2013 but Not in 2014

**Description:** This query retrieves products that received customer reviews in the year 2013 but did not receive any reviews in 2014. It uses the `EXCEPT` set operator to identify these products.

**Inputs:**

- `ProductID`: The unique identifier for each product.
- `ReviewDate`: The date when the review was written.

**Outputs:**

- `ProductID`: The unique identifier for the product.
- `ProductName`: The name of the product.

**Steps:**

1. Select distinct products that have reviews dated in 2013.
2. Select distinct products that have reviews dated in 2014.
3. Use `EXCEPT` to find products reviewed in 2013 but not in 2014.

**Assumptions:**

- The data is available in the `Production.Product` and `Production.ProductReview` tables.
- Products have unique `ProductID`s in the `Production.Product` table.

**Example Output**

| ProductID | ProductName |
| --- | --- |
| 680 | HL Road Frame - Black, 58 |
| 712 | Sport-100 Helmet, Red |

In [85]:
USE AdventureWorks2019
SELECT DISTINCT
    p.ProductID,
    p.Name AS ProductName
FROM
    Production.Product AS p
INNER JOIN
    Production.ProductReview AS pr ON p.ProductID = pr.ProductID
WHERE
    YEAR(pr.ReviewDate) = 2013

EXCEPT

SELECT DISTINCT
    p.ProductID,
    p.Name AS ProductName
FROM
    Production.Product AS p
INNER JOIN
    Production.ProductReview AS pr ON p.ProductID = pr.ProductID
WHERE
    YEAR(pr.ReviewDate) = 2014;


(3 rows affected)

Total execution time: 00:00:00.003

ProductID ProductName 709 Mountain Bike Socks, M 798 Road-550-W Yellow, 40 937 HL Mountain Pedal